src/main/java/com/google/api/gax/retrying/RetryingFutureImpl.java

```diff
@@ -0,0 +1,183 @@
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.api.gax.retrying;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.gax.core.ApiFutureCallback;
+import com.google.common.util.concurrent.AbstractFuture;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Future;
+
+/**
+ * For internal use only.
+ *
+ * <p>
+ * This class is the key component of the retry logic. It implements the {@link RetryingFuture}
+ * facade interface, and does the following:
+ *
+ * <ul>
+ * <li>Schedules the next attempt in case of a failure using the callback chaining technique.
+ * <li>Terminates retrying process if no more retries are accepted.
+ * <li>Propagates future cancellation in both directions (from this to the attempt and from the
+ * attempt to this)
+ * </ul>
+ *
+ * This class is thread-safe.
+ */
+class RetryingFutureImpl<ResponseT> extends AbstractFuture<ResponseT>
+    implements RetryingFuture<ResponseT> {
+
+  private final Object lock = new Object();
+  private final Callable<ResponseT> callable;
+
+  private final RetryAlgorithm retryAlgorithm;
+  private final RetryingExecutor<ResponseT> retryingExecutor;
+
+  private volatile TimedAttemptSettings attemptSettings;
+  private volatile AttemptFutureCallback attemptFutureCallback;
+
+  RetryingFutureImpl(
+      Callable<ResponseT> callable,
+      RetryAlgorithm retryAlgorithm,
+      RetryingExecutor<ResponseT> retryingExecutor) {
+    this.callable = checkNotNull(callable);
+    this.retryAlgorithm = checkNotNull(retryAlgorithm);
+    this.retryingExecutor = checkNotNull(retryingExecutor);
+
+    this.attemptSettings = retryAlgorithm.createFirstAttempt();
+  }
+
+  @Override
+  public boolean cancel(boolean mayInterruptIfRunning) {
+    synchronized (lock) {
+      if (attemptFutureCallback != null) {
+        if (attemptFutureCallback.attemptFuture.cancel(mayInterruptIfRunning)) {
+          super.cancel(mayInterruptIfRunning);
+        }
+        return isCancelled();
+      } else {
+        return super.cancel(mayInterruptIfRunning);
+      }
+    }
+  }
+
+  @Override
+  public void setAttemptFuture(Future<ResponseT> attemptFuture) {
+    if (isDone()) {
+      return;
+    }
+    synchronized (lock) {
+      if (isDone()) {
+        return;
+      }
+      if (attemptFuture != null) {
+        attemptFutureCallback = new AttemptFutureCallback(attemptFuture);
+        Futures.addCallback((ListenableFuture) attemptFuture, attemptFutureCallback);
```

Comment:
You require a `ListenableFuture` to be passed into `setAttemptFuture` (otherwise there will be a `ClassCastException`), but this is not documented, and additionally, it requires usage of Guava to call this method. `ApiFuture` was designed to fix this problem: no casting is necessary, and it provides the functionality that people need to interact with this library. I am not personally concerned about the extra layers of wrapping.